warl_proposal/csr_template_draft1.yaml

```diff
@@ -0,0 +1,130 @@
+############################################ Template description ############################################
+# csr template for csrs with subfields
+<name>:                                   # name of the csr
+  description: <text>                     # textual description of the csr
+  address: <hex>                          # address of the CSR
+  priv_mode: <D/M/H/S/U>                  # privilege mode that owns the register
+  reset-val: <hex>                        # Reset value of the register. This an accumulation of the all reset values of the 
+                                          # sub-field 
+  rv32:                                   # this node and its subsequent fields can exist if [M/S/U]XL value can be 1
+    exists: <boolean>                     # indicates if the csr exists in rv32 mode or not. When False, all fields below 
+                                          # will be trimmed off in the checked yaml. False also indicates that 
+                                          # access-exception should be generated
+    subfields:                            # a quick summary of the list of all subfields of the csr including a list of 
```

Comment:
Why subfield is needed here? Seems like it's redundant as all the fields are listed in details later.

Reply:
This serves as a list of all the subfields that this csr could have along with the WPRI fields. An implementation may choose to not implement certain subfields of this csr which will be captured in the subfield definition itself.

This table is like the reference of all fields available within this csr. This would typically be predefined for each csr within the checker and user would not be burdened to describe this. We could then check if all subfields have been captured by the implementation or not (even if they are not implemented the user has to explicitly define them as not-implemented)

Reply:
This serves as a list of all the subfields that this csr could have along with the WPRI fields. An implementation may choose to not implement certain subfields of this csr which will be captured in the subfield definition itself.

This table is like the reference of all fields available within this csr. This would typically be predefined for each csr within the checker and user would not be burdened to describe this. We could then check if all subfields have been captured by the implementation or not (even if they are not implemented the user has to explicitly define them as not-implemented)

Reply:
If it's predefined in the checker based on the spec, should it be removed from this format? I feel that we should have a template with all the fields listed in detail below. User could change some to unimplemented, and the checker should check if any mandatory field is missing.

Reply:
I probably should have mentioned this in the begining: the file csr_template_draft1.yaml is what the checker would generate at the end. The user provided input yaml need not have all the informations that you see in this file. 

Minimally the user should specify the WARL funcs atleast.. most of the other fields like : priv_mode, msb, lsb,e tc would be taken care of by the checker since they are fixed by the spec and the user shouldn't be burdened with re-defining them.

But, the user can also choose to re-define all the fields but they would fail certain checks of the checker. This allows a user to define custom-csrs in the same template as well.

Reply:
Cool, this makes perfect sense to me.

Reply:
By the way, I think later in the user doc, there should be some example of use model with minimum user input.

Reply:
absolutely. This was done to get quick feedback from people. Will enhance for better understanding. Thanks.

BTW, if you have something defined for the Ibex core, I could use that as an example itself?

Reply:
This is format we use to generate CSR test
https://github.com/lowRISC/ibex/blob/master/dv/uvm/riscv_dv_extension/csr_description.yaml
We don't have range/constraint/shadow yet. We plan to add range shortly. It will be great to converge to one format and share the effort.